keps/sig-api-machinery/2309-idl/syntax.md

```diff
@@ -0,0 +1,635 @@
+   corresponding concepts in the language.  For instance:
+
+   * Quantity is fundamentally a primitive in Kubernetes, and thus should
+     have a corresponding primitive in our IDL.
+
+   * Kind is a fundamental concept in Kubernetes, so it should just be
+     possible to declare a kind in the IDL.
+
+2. Kubernetes API requirements should be encoded in the language:
+   disallowed constructs (and many bad practices) in Kubernetes APIs
+   should not be possible to represent in the IDL.
+
+3. Kubernetes API recommendations should be the default behavior in the
+   language: when not explicitly overridden, users should get our
+   recommended behavior by default:
+
+   * For example, the recommendation is new list-maps have a merge key of
+     name.  When merge keys aren’t manually specified, this is the default
+     one.
+
+4. All information conveyed should be as structured as possible
+
+5. Naming in the IDL describes the serialized form.  Language bindings
+   (Go) are derived automatically from that.  Specific overrides for
+   legacy codebases occur via markers, since they are exceptions to the
+   rules.
+
+## Examples
+
+***Author's note***: the cannonical source for these examples is [in the
+repository][repo-example]; the example reproduced here is just to make it
+easier to comment.
+
+[repo-example]: https://github.com/DirectXMan12/idl/blob/main/kdlc/testdata/all-around.kdl
+
+```
+// double slashes add comments
+/* same with block comments */
+
+// Triple-slash comments denote documentation
+// If a line contains `# Words`, the following lines
+// will be placed in the corresponding OpenAPI documentation
+// field.
+
+/// core/v1 describes core k8s concepts
+group-version(group:"core", version:"v1") {
+    /// Pod is a group of related, colocated processes
+    ///
+    /// # Example
+    /// {"apiVersion": "v1", "kind": "Pod", "spec": {}}
+    kind Pod {
+        // types may be nested -- this is just a convenience for namespacing
+        // it does not imply privacy or anything
+
+        spec: Spec,
+        struct Spec {
+            // list-maps are ordered maps whose keys appear in their body.
+            // not specifying a key defaults to `key: name`.
+            // Go represents them as lists with specific merge keys, other
+            // languages may represent them other ways.
+            // They are automatically merged as maps in SMD.
+
+            // optional fields must list the keyword `optional` before the
+            // type.
+
+            /// list of volumes that can be mounted by containers belonging to
+            /// the pod
+            volumes: optional list-map(value: Volume),
+
+            // ...
+
+            dnsPolicy: optional(default: ClusterFirst) DNSPolicy,
+
+            // simple-maps are unordered maps from string-type values to
+            // primitive values
+            nodeSelector: optional simple-map(value: string),
```

Comment:
is there a connection to atomic?

Reply:
hmm... there's currently not a knob for configuring the atomicity of maps.  tbh, I've never seen it used anywhere, but we could certainly add it.  For the list-y types, their atomicity is determined by the type in use (list/set/list-map is default/set/map).

Reply:
Oh, it's used.  We definitiely need to keep it.  I'm not sure it is inferable, either

Reply:
o_O where?

I've found the ones for list, but

```
~[go:kubernetes] $ rg '\+mapType'
```

doesn't return anything 

Found one example for structType (metav1.LabelSelector), so we'll want that at the very least (I'm fine adding both though -- just couldn't find a use of `mapType` in k/k or the wild afaict).

Reply:
Lists are atomic by default, but we'd rather make that field required. That's why the "exception" file is full of lists not specifying a `mapType`.